File: 2025-10-October-GFG-POTD/October 30 - Replace O's with X's/Solution.java
class Solution {
    public void fill(char[][] grid) {
        if (grid == null || grid.length == 0) return;
        
        int n = grid.length;
        int m = grid[0].length;
        
        // Step 1: Mark 'O's connected to border
        for (int i = 0; i < n; i++) {
            if (grid[i][0] == 'O') dfs(grid, i, 0, n, m);
            if (grid[i][m - 1] == 'O') dfs(grid, i, m - 1, n, m);
        }
        for (int j = 0; j < m; j++) {
            if (grid[0][j] == 'O') dfs(grid, 0, j, n, m);
            if (grid[n - 1][j] == 'O') dfs(grid, n - 1, j, n, m);
        }

        // Step 2: Flip remaining 'O' to 'X' and '#' back to 'O'
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 'O') grid[i][j] = 'X';
                else if (grid[i][j] == '#') grid[i][j] = 'O';
            }
        }
    }

    // DFS helper to mark safe 'O's
    private void dfs(char[][] grid, int i, int j, int n, int m) {
        if (i < 0 || i >= n || j < 0 || j >= m || grid[i][j] != 'O') return;
        
        grid[i][j] = '#'; // mark as visited (safe)
        
        dfs(grid, i + 1, j, n, m);
        dfs(grid, i - 1, j, n, m);
        dfs(grid, i, j + 1, n, m);
        dfs(grid, i, j - 1, n, m);
    }
}

File: 2025-10-October-GFG-POTD/October 30 - Replace O's with X's/Solution.py
class Solution:
    def fill(self, grid):
        if not grid or not grid[0]:
            return
        
        n, m = len(grid), len(grid[0])
        
        # Helper DFS function
        def dfs(i, j):
            if i < 0 or i >= n or j < 0 or j >= m or grid[i][j] != 'O':
                return
            grid[i][j] = '#'  # mark as safe
            dfs(i + 1, j)
            dfs(i - 1, j)
            dfs(i, j + 1)
            dfs(i, j - 1)
        
        # Step 1: Mark border-connected 'O's
        for i in range(n):
            if grid[i][0] == 'O':
                dfs(i, 0)
            if grid[i][m - 1] == 'O':
                dfs(i, m - 1)
        for j in range(m):
            if grid[0][j] == 'O':
                dfs(0, j)
            if grid[n - 1][j] == 'O':
                dfs(n - 1, j)
        
        # Step 2: Replace surrounded 'O' → 'X' and '#' → 'O'
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 'O':
                    grid[i][j] = 'X'
                elif grid[i][j] == '#':
                    grid[i][j] = 'O'
